functorch/_src/eager_transforms.py

```diff
@@ -1076,3 +1082,47 @@ def wrapper(*args, **kwargs):
         grad, _ = results
         return grad
     return wrapper
+
+
+def assert_wrapped_functional(
+        unwrapped_tensor_inputs: List[torch.Tensor],
+        wrapped_tensor_inputs: List[torch.Tensor]) -> None:
+    if len(unwrapped_tensor_inputs) != len(wrapped_tensor_inputs):
+        raise Exception()
+    for i in range(len(unwrapped_tensor_inputs)):
+        _assert_wrapped_functional(unwrapped_tensor_inputs[i], wrapped_tensor_inputs[i])
+
+
+def propagate_functional_input_mutations(
+        unwrapped_tensor_inputs: List[torch.Tensor],
+        wrapped_tensor_inputs: List[torch.Tensor]) -> None:
+    if len(unwrapped_tensor_inputs) != len(wrapped_tensor_inputs):
+        raise Exception()
+    for i in range(len(unwrapped_tensor_inputs)):
+        _propagate_functional_input_mutation(unwrapped_tensor_inputs[i], wrapped_tensor_inputs[i])
+
+
+def functionalize(func: Callable) -> Callable:
+    @wraps(func)
+    def wrapped(*args, **kwargs):
+        try:
+            func_level = _func_increment_nesting()
+            func_args = [_wrap_functional_tensor(x, func_level) if isinstance(x, torch.Tensor) else x for x in args]
+
+            unwrapped_tensor_inputs = [t for t in args if isinstance(t, torch.Tensor)]
+            wrapped_tensor_inputs = [t for t in func_args if isinstance(t, torch.Tensor)]
+            assert_wrapped_functional(unwrapped_tensor_inputs, wrapped_tensor_inputs)
+
+            func_outputs = func(*func_args, **kwargs)
+            flattened_outputs, _ = tree_flatten(func_outputs)
+            for a in func_args:
+                if isinstance(a, torch.Tensor):
+                    # Call sync_() on the inputs, to ensure that they still get mutated inplace if the original
+                    # program mutated its inputs
+                    torch._sync(a)
+
+            propagate_functional_input_mutations(unwrapped_tensor_inputs, wrapped_tensor_inputs)
+            return [_unwrap_functional_tensor(x) if isinstance(x, torch.Tensor) else x for x in flattened_outputs]
```

Comment:
If the original function returns a single tensor output, then functionalize(f) is going to return a List, right?

The way to get this to return something of the same python structure as func_outputs is to do a tree_unflatten using the spec obtained from `flattened_outputs, spec = tree_flatten(func_outputs)`

Or you can just `output = tree_map(something_here, func_outputs)` where something_here is `_unwrap_functional_tensor(x) if isinstance(x, torch.Tensor) else x`